src/main.rs

```diff
@@ -342,3 +307,75 @@ fn err_to_warp(e: podsync::Error) -> impl warp::Reply {
 pub struct QuerySince {
     since: crate::time::Timestamp,
 }
+
+#[derive(Copy, Clone, Debug, PartialEq, Eq)]
+enum UsernameFormat {
+    Name,
+    NameJson,
+}
+impl UsernameFormat {
+    pub fn convert<'a>(&self, username: &'a str) -> Result<&'a str, warp::Rejection> {
+        match self {
+            Self::Name => Ok(username),
+            Self::NameJson => split_format_json(username).map_err(warp::reject::custom),
+        }
+    }
+}
+
+fn cookie_authorize(
+    username_fmt: UsernameFormat,
+    podsync: Arc<PodSync>,
+) -> impl Filter<Extract = (PodSyncAuthed<true>,), Error = warp::Rejection> + Clone {
+    let cookie_auth_check = warp::cookie(COOKIE_NAME).and_then({
+        move |session_id: SessionId| {
+            let podsync = Arc::clone(&podsync);
+
+            async move {
+                podsync
+                    .authenticate(session_id)
+                    .await
+                    .map_err(warp::reject::custom)
+            }
+        }
+    });
+
+    warp::path::param::<String>()
+        .and(cookie_auth_check)
+        .and_then(move |username: String, podsync: PodSyncAuthed| async move {
+            podsync
+                .with_user(username_fmt.convert(&username)?)
+                .map_err(warp::reject::custom)
```

Comment:
In this case, we'll want a hard error rather than something recoverable, since the client has passed a cookie check but with the wrong username, which we want to indicate to the client (unlike the optimistic check for a cookie above).

I'm thinking of doing a little overhaul of the rejections to make rejecting mid-filter easier (avoiding my slightly hacky `result_to_ok` business) - I'll leave it up to you if you want a go at returning a [4xx error](https://github.com/bobrippling/podsync/blob/6889de9/src/podsync.rs#L44) here or if not, I'll sweep it up in my overhaul.